### DISCUSSION TITLE
How to use RGCNConv and NNConv

### DISCUSSION THREAD

**Discussion by vlongle** [pyg-team/pytorch_geometric#2968] [Q&A] [answered]
Hi,

I'm working on a grid-world problem with connected rooms. Each room is represented as a node and some doors between rooms might be blocked. As such, my data is
```
Data(edge_attr=[58, 2], edge_index=[2, 58], edge_type=[58], x=[24, 4])
```
where the `edge_type` indicates whether the edge is blocked and `edge_attr` has the following edge feature (`blocked`, `number_of_doors`).  I tried to use RGCNConv and NNConv following `pytorch_geometric/examples` as follows
![Screen Shot 2021-08-13 at 12 20 58 AM](https://user-images.githubusercontent.com/37346413/129304439-0ffd5c88-764c-4eca-af5e-9ee8a191ee41.png)
but neither of them worked. The RGCN returned this error
![Screen Shot 2021-08-13 at 12 23 09 AM](https://user-images.githubusercontent.com/37346413/129304576-ac8d11de-6e10-497f-a540-111ab7e2a13d.png)
and NNConv this
![Screen Shot 2021-08-13 at 12 23 53 AM](https://user-images.githubusercontent.com/37346413/129304663-97db85f2-75e3-4eb9-aa27-1e627928c480.png)
Any help would be much appreciated!

**Comment by rusty1s** [comment, marked as answer]
Not really sure yet why the `RGCNConv` example does not work for you (can you create some custom data to reproduce?). In the second example, the `nn` needs to map to `in_channels * out_channels` of the `NNConv`:
```python
nn1 = Sequential(Linear(num_edge_features, num_node_features * hidden_dim), ReLU()
self.conv1 = NNConv(num_node_features, hidden_dim, nn1, aggr='mean')
nn2 = Sequential(Linear(num_edge_features, hidden_dim * hidden_dim), ReLU()
self.conv2 = NNConv(hidden_dim, hidden_dim, nn1, aggr='mean')
```

**Reply by vlongle** [reply]
Here is a small example where both of my classes did not work
![Screen Shot 2021-08-13 at 10 01 10 AM](https://user-images.githubusercontent.com/37346413/129368771-144092f5-0ef1-4e87-a3e7-96cc9f33963f.png)

For the R-GCN, the error was 
![Screen Shot 2021-08-13 at 10 02 15 AM](https://user-images.githubusercontent.com/37346413/129368861-f84e4714-d033-49ee-985a-de43b7a4e636.png)
and for the NN-Conv (after I have added your fix)
![Screen Shot 2021-08-13 at 10 02 52 AM](https://user-images.githubusercontent.com/37346413/129368939-621f2496-35e5-4982-83c1-cf36125dda43.png)

Also, for the NNConv, I saw in the documentation that `nn` needs to output `in_channels * out_channels` but the example https://github.com/rusty1s/pytorch_geometric/blob/master/examples/mnist_nn_conv.py#L31 has `nn1` outputs 32 which is also the out_channels in `self.conv1` instead of having `nn1` output `d.num_features * 32` so that's why I was a bit confused.

Thanks!

**Reply by rusty1s** [reply]
I modified the examples in `mnist_nn_conv.py` to make that more clear. Thanks for pointing out.

Regarding your issues with `RGCNConv`: `data.x` should be of type `float`. With this, the following example works for me:
```python
import networkx as nx
from torch_geometric.utils import from_networkx
from torch_geometric.nn import RGCNConv

G = nx.Graph()
G.add_edge(1, 2)
G.nodes[1]['x'] = [0., 0., 0.]
G.nodes[2]['x'] = [0., 1., 0.]

G.edges[(1, 2)]['edge_attr'] = [1., 0., 10., 5.]
G.edges[(1, 2)]['edge_type'] = 0

data = from_networkx(G)

conv = RGCNConv(in_channels=3, out_channels=6, num_relations=1)
out = conv(data.x, data.edge_index, data.edge_type)
print(out.shape)
```
